Persister

Lukas0907 opened the pull request:
This PR ~builds on the previous work in the redis branch (PR #1) and~ introduces the concept of a *Persister* which abstracts away the current `JOBDIR` storage mechanism. The basic idea is that instead of checking if the `JOBDIR` setting is configured and using it as a directory where arbitrary data can be read from and written to, a Persister should be used instead. A Persister offers a common interface to its users so they don‘t have to worry how the data is persisted.

Currently there ~are two implementations~ is only one implementation: A disk-based persister (`scrapy.utils.job.DiskPersister`) which basically mimics the current behavior in a backwards-compatible way. ~The second persister is the `scrapy.utils.job.RedisPersister` which uses Redis to persist data.~ The concrete implementaton of the persister can be configured via the commonly used dependency injection mechanism: The `JOB_PERSISTER` setting expects a fully-qualified name of a class that implements the persister interface.

Lukas0907 committed:
Introduce Persister interface

Gallaecio commented on scrapy/core/scheduler.py:
```diff
@@ -39,10 +38,10 @@ class Scheduler:
     Also, it handles dupefilters.
     """
-    def __init__(self, dupefilter, jobdir=None, dqclass=None, mqclass=None,
+    def __init__(self, dupefilter, persister=None, dqclass=None, mqclass=None,
```
This is a backward-incompatible change.

The backward-compatible approach would be to:
- Add `persister=None` as the last keyword argument.
- If `jobdir` is not `None`:
  - [Issue a deprecation warning](https://docs.python.org/3/library/warnings.html#warnings.warn). Find usages of `ScrapyDeprecationWarning` for an example. Choosing the right `stacklevel` may require some thinking.
  - Handle `jobdir` in a backward-compatible way.

Gallaecio commented on scrapy/core/scheduler.py:
```diff
@@ -39,10 +38,10 @@ class Scheduler:
+    def __init__(self, dupefilter, persister=None, dqclass=None, mqclass=None,
                  logunser=False, stats=None, pqclass=None, crawler=None):
         self.df = dupefilter
-        self.dqdir = self._dqdir(jobdir)
```
This is also a backward-incompatible change. To keep things backward-compatible, you would need to transform this attribute into a property, and when it is read, you should issue a deprecation warning. Otherwise, it should behave the same as before.

Lukas0907 replied:
Thanks, changed: https://github.com/Lukas0907/scrapy/pull/2/files#diff-1e8b9699ff2fb363b837e0da1b7eb876R126-R136

Gallaecio commented on scrapy/core/scheduler.py:
```diff
@@ -150,33 +154,29 @@ def _mq(self):
 
     def _dq(self):
         """ Create a new priority queue instance, with disk storage """
-        state = self._read_dqs_state(self.dqdir)
+        state = self._read_dqs_state(self.persister)
```
:lipstick: The issue comes from the original code, but I believe you can refactor the method not to use a parameter at all.

Gallaecio commented on scrapy/core/scheduler.py:
```diff
@@ -150,33 +154,29 @@ def _mq(self):
-    def _read_dqs_state(self, dqdir):
-        path = join(dqdir, 'active.json')
-        if not exists(path):
+    def _persist_key(self, key):
```
:lipstick: This method is always called with the same argument, maybe some refactoring makes sense.

Gallaecio commented on scrapy/core/scheduler.py:
```diff
@@ -150,33 +154,29 @@ def _mq(self):
+    def _write_dqs_state(self, persister, state):
+        persister.set(
+            self._persist_key('active.json'), json.dumps(state).encode("utf8")
+        )
```
I’m thinking that maybe the `Persister` class should be renamed as a class that is more specific to the purpose that is has in Scrapy, e.g. `PersistentRequestQueue`, and provide public methods that `_read_dqs_state`  and `_write_dqs_state`, e.g. `get_state` and `set_state`.

Lukas0907 replied:
The JOBDIR is currently used to persist spider state, request fingerprints and the disk queues. The idea of the persister is therefore to be used in all of these three places. Therefore I was thinking about about renaming it to `JobPersister` because it's not really specific to request queues.

ejulio reviewed (commented):
Isn't this a "big change in scrapy"?
If so, I'd break it in a PR that we could send to scrapy itself, and another one that introduces Redis stuff.
Otherwise, we'll need to do it later, by proceeding to a new library

whalebot-helmsman commented:
> Isn't this a "big change in scrapy"?

Yes, I am also think it is very big change. We should discuss less invasive ways to reach what is needed.

Lukas0907 commented:
I rebased the PR on current `master` and removed the Redis-specific parts. I also moved the `persister` argument to the end and reintroduced the `jobdir`/`path` arguments to preserve backwards-compatibility (if the arguments are set, a warning is issued).

ejulio commented on scrapy/extensions/spiderstate.py:
```diff
@@ -1,40 +1,52 @@
-            with open(self.statefn, 'wb') as f:
-                pickle.dump(spider.state, f, protocol=4)
+        if self.persister:
+            self.persister.set('spider.state', pickle.dumps(spider.state, protocol=4))
```
I'm not sure if the dumping code should be here.
Looking in terms of responsibility, it seems a bit more related to the persister.
Basically, it should receive some value and know how to dump (encode) and load (decode) it.

Lukas0907 replied:
I agree that the persister/storage/repository should take care of how to serialize the data. However, there are currently three different ways how data is serialized:

- pickle
- JSON
- line-separated text

I didn't want to make a backwards incompatible change so I left it for the caller to decide.

ejulio commented on scrapy/utils/job.py:
```diff
@@ -1,8 +1,99 @@
+def persister_from_settings(settings):
+    persister_cls = settings['JOB_PERSISTER']
+    if not persister_cls:
+        return None
```
I guess this is fine, in a transition period, but then we should raise an exception if is not set, or if it is not a valid class.
Otherwise, the crawling process would run and it'll fail later, at the time that some component interacts with the persister.

ejulio replied:
Ideally, we should always aim to fail fast

Lukas0907 replied:
By default it is set to `scrapy.utils.job.DiskPersister` so it would require somebody to purposefully set it to `None`. Do you think this is a problem even considering the default setting?

ejulio replied:
My main point is if it's invalid (wrong path name, as it can be common, since we're dealing with strings), when is it going to fail?
And, if it fails, how will the user know about it?

ejulio reviewed (commented):
Changes look good.
Just a case for naming.
I'd vote for `Storage` instead of `Persister`, with methods `store` and `retrieve`. Bu that's open for debate.
Another option is to consider it a `Repository` and use that pattern, which is basically an abstraction for a place to store data.
So, both `Storage` and `Repository` are fine for me

Lukas0907 commented:
>  So, both `Storage` and `Repository` are fine for me

Fine by me, too, of course!
